upki/src/main.rs

```diff
@@ -18,21 +19,44 @@ async fn main() -> Result<(), Report> {
             .init();
     }
 
-    let local = match args.cache_dir {
-        Some(dir) => dir,
-        None => match directories::ProjectDirs::from("dev", "rustls", "upki") {
-            Some(dirs) => dirs.data_local_dir().to_owned(),
-            None => return Err(eyre!("cannot determine home directory")),
-        },
+    let maybe_config_path = match args.config_file {
+        Some(path) => Some(path),
+        None => Config::find_config_file().wrap_err("cannot find configuration path")?,
     };
-    tracing::info!("local directory is {local:?}");
 
-    match args.command {
-        Command::Fetch { url, dry_run } => fetch::fetch(&local, &url, dry_run).await?,
-        Command::Verify => fetch::verify(&local)?,
-        Command::ShowCacheDir => println!("{}", local.display()),
+    if let Command::ShowConfigPath = args.command {
+        let config_path = match maybe_config_path {
+            Some(path) => path,
+            None => Config::create_preferred_config_path()?,
+        };
+        println!("{}", config_path.display());
+        return Ok(());
     }
 
+    let config = match maybe_config_path {
+        Some(config_path) => Config::load_from_file(&config_path)
+            .wrap_err_with(|| format!("cannot load configuration file from {config_path:?}"))?,
+        None => Config::try_default()?,
+    };
```

Comment:
This sequence of stuff seems somewhat confusing. I think maybe `Config::new()` (replacing `try_default()`) should yield `Result<(Self, PathBuf), _>` so that we always have both a config (potentially with default values only) and the path to a config (which may not exist).